src/server/endpoints/admin.py

```diff
@@ -67,44 +59,50 @@ def removal_request_redirect():
 def _index():
     token = _require_admin()
     flags = dict()
-    if request.method == "POST":
-        # register a new user
-        if not user_exists(user_email=request.values["email"], api_key=request.values["api_key"]):
-            User.create_user(
-                request.values["api_key"],
-                request.values["email"],
-                _parse_roles(request.values.getlist("roles")),
-            )
-            flags["banner"] = "Successfully Added"
-        else:
-            flags["banner"] = "User with such email and/or api key already exists."
-    users = [user.as_dict for user in User.list_users()]
-    return _render("overview", token, flags, users=users, user=dict())
+    with WriteSession() as session:
+        if request.method == "POST":
+            # register a new user
+            if not User.find_user(
+                    user_email=request.values["email"], api_key=request.values["api_key"],
+                    session=session):
+                User.create_user(
+                    api_key=request.values["api_key"],
+                    email=request.values["email"],
+                    user_roles=set(request.values.getlist("roles")),
+                    session=session
+                )
+                flags["banner"] = "Successfully Added"
+            else:
+                flags["banner"] = "User with such email and/or api key already exists."
+        users = [user.as_dict for user in session.query(User).all()]
+        return _render("overview", token, flags, session=session, users=users, user=dict())
 
 
 @bp.route("/<int:user_id>", methods=["GET", "PUT", "POST", "DELETE"])
 def _detail(user_id: int):
     token = _require_admin()
-    user = User.find_user(user_id=user_id)
-    if not user:
-        raise NotFound()
-    if request.method == "DELETE" or "delete" in request.values:
-        User.delete_user(user.id)
-        return redirect(f"./?auth={token}")
-    flags = dict()
-    if request.method in ["PUT", "POST"]:
-        user_check = User.find_user(api_key=request.values["api_key"], user_email=request.values["email"])
-        if user_check and user_check.id != user.id:
-            flags["banner"] = "Could not update user; same api_key and/or email already exists."
-        else:
-            user = user.update_user(
-                user=user,
-                api_key=request.values["api_key"],
-                email=request.values["email"],
-                roles=_parse_roles(request.values.getlist("roles")),
-            )
-            flags["banner"] = "Successfully Saved"
-    return _render("detail", token, flags, user=user.as_dict)
+    with WriteSession() as session:
```

Comment:
using this context manager lets us do calls from `models.py:User` (`find_user()`, `delete_user()`, etc) within a single deliberate session instead of letting them use their own individual default sessions.  it also ensures that all of those calls happen on the session thats attached to the primary db server to eliminate sync issues (`find_user` defaults to the [load-balanced/replicated/read]Session and the other methods modify user objects with their default WriteSessions).

we **_could_** probably put a `@default_session(WriteSession)` on the `_detail()` method and remove this `with`, but i like doing the `_require_admin()` call outside of setting up the session (because it will raise and exception and bail early if the request isnt auth'd).  plus it might require some finagling if `@default_session` has any interactions with the other decorator (`@bp.route`) thats already there.

Reply:
> i like doing the _require_admin() call outside of setting up the session

good enough for me!